src/libstd/path.rs

```diff
@@ -0,0 +1,2415 @@
+                self.path = &self.path[1..];
+                return Some(Component::RootDir)
+            } else if let Some(p) = self.prefix {
+                if p.has_implicit_root() && !p.is_verbatim() {
+                    return Some(Component::RootDir)
+                }
+            }
+            self.next()
+        } else if self.front == State::Body {
+            while !self.path.is_empty() {
+                let (size, comp) = self.parse_next_component();
+                self.path = &self.path[size ..];
+                if comp.is_some() { return comp }
+            }
+            self.front = State::Suffix;
+            self.next()
+        } else { // self.front == State::Suffix
+            self.front = State::Done;
+            if self.prefix_verbatim() {
+                Some(Component::Empty)
+            } else {
+                Some(Component::CurDir)
+            }
+        }
+    }
+}
+
+impl<'a> DoubleEndedIterator for Components<'a> {
+    fn next_back(&mut self) -> Option<Component<'a>> {
+        if self.finished() {
+            None
+        } else if self.back == State::Suffix {
+            self.back = State::Body;
+            if self.prefix_verbatim() {
+                Some(Component::Empty)
+            } else {
+                Some(Component::CurDir)
+            }
+        } else if self.back == State::Body {
+            while self.path.len() > self.prefix_and_root() {
+                let (size, comp) = self.parse_next_component_back();
+                self.path = &self.path[.. self.path.len() - size];
+                if comp.is_some() { return comp }
+            }
+            self.back = State::Root;
+            self.next_back()
+        } else if self.back == State::Root {
+            self.back = State::Prefix;
+            if self.has_physical_root {
+                self.path = &self.path[.. self.path.len() - 1];
+                return Some(Component::RootDir)
+            } else if let Some(p) = self.prefix {
+                if p.has_implicit_root() && !p.is_verbatim() {
+                    return Some(Component::RootDir)
+                }
+            }
+            self.next_back()
+        } else { // self.back == State::Prefix
+            self.back = State::Done;
+            if self.prefix_len() > 0 {
+                //debug_assert!(self.path.len() == self.prefix.len());
+                Some(Component::Prefix(unsafe { Path::from_u8_slice(self.path) }))
+            } else {
+                None
+            }
+        }
+    }
+}
+
+fn optional_path(path: &Path) -> Option<&Path> {
+    if path.as_u8_slice().is_empty() { None } else { Some(path) }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Basic types and traits
+////////////////////////////////////////////////////////////////////////////////
+
+/// An owned, mutable path (akin to `String`).
+pub struct PathBuf {
+    inner: OsString
+}
+
+/// A slice of a path (akin to `str`).
```

Comment:
This may want to explain how it's a DST and perhaps show some common methods of construction/slicing/etc.

I suspect that this will be one of the first landing pages for clicking on `Path` documentation in rustdoc.